src/IECoreScene/MeshAlgoSplit.cpp

```diff
@@ -0,0 +1,545 @@
+	// original id corresponding to each id of the output
+	void getDataRemapping( std::vector<int> &dataRemap )
+	{
+		dataRemap.resize( m_numIdsUsed );
+		for( unsigned int i = 0; i < m_fromOldIds.size(); i++ )
+		{
+			auto &blockPointer = m_fromOldIds[ i ];
+			if( blockPointer )
+			{
+				std::vector<int> &block = *blockPointer;
+				for( int j = 0; j < m_blockSize; j++ )
+				{
+					int newId = block[j];
+					if( newId != -1 )
+					{
+						int oldId = i * m_blockSize + j;
+						dataRemap[ newId ] = oldId;
+					}
+				}
+			}
+		}
+	}
+
+private:
+	// IntVectorData to hold the new indices
+	IntVectorDataPtr m_newIndicesData;
+	std::vector< int > &m_newIndices;
+
+	// A performance tuning value determining how large the blocks that are allocated to hold ids are.
+	const int m_blockSize;
+
+	// Store the mapping from old ids to new ids.  The outer vector holds a unique_ptr for each
+	// block of m_blockSize ids in the original id range.  These pointers are null if no ids from
+	// that block have been used.  Once a block is used, it is allocated with a vector that is set
+	// to -1 for ids which have not been used, and the new corresponding id for ids which have been
+	// used
+	std::vector< std::unique_ptr< std::vector< int > > > m_fromOldIds;
+
+	// How many unique ids have appeared in the indices added so far
+	int m_numIdsUsed;
+
+};
+
+struct ResamplePrimitiveVariableFunctor
+{
+
+	// Resample a primitive variable, given all the necessary indices from the mesh.
+	//
+	// The implementation is all pretty straightforward, since we already have the
+	// indices we need for each case prepared, and we can use Reindexer to deal with
+	// indexed primvars.
+	//
+	// All the bogus [[maybe_unused]] statements are required due to this bug:
+	// https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81676
+	// which I think was fixed in GCC 10, but I haven't tested
+
+	template<typename T>
+	PrimitiveVariable operator()( [[maybe_unused]] const T *data, const PrimitiveVariable &primVar, [[maybe_unused]] int startIndex, [[maybe_unused]] int numFaces, [[maybe_unused]] int totalFaceVerts, [[maybe_unused]] const std::vector<int> &faceRemap, [[maybe_unused]] const std::vector<int> &verticesPerFace, [[maybe_unused]] const std::vector<int> &faceIndices, [[maybe_unused]] const std::vector<int> &vertRemapBackwards, [[maybe_unused]] const Canceller *canceller )
+	{
+		if(
+			primVar.interpolation != PrimitiveVariable::Interpolation::Uniform &&
+			primVar.interpolation != PrimitiveVariable::Interpolation::Vertex &&
+			primVar.interpolation != PrimitiveVariable::Interpolation::Varying &&
+			primVar.interpolation != PrimitiveVariable::Interpolation::FaceVarying
+		)
+		{
+			// Just copying works for constants
+			return primVar;
+		}
+
+		if constexpr ( ! IECore::TypeTraits::IsVectorTypedData< T >::value )
+		{
+			throw IECore::Exception( "Malformed PrimitiveVariable, this shouldn't be possible." );
+		}
+		else
```

Comment:
Since the `if()` above throws, we could omit this `else` and remove an extra level of indentation for the main contents of this function.